A55-stateful-session-affinity.md

```diff
@@ -0,0 +1,348 @@
+A55: Stateful Session Affinity for Proxyless gRPC
```

Comment:
I agree that the idea of cookie based session affinity has no xds dependency. But making this usable without xds means abstracting out http-filter and CDS dependencies and that's non-trivial. So I will just make the required changes to indicate "xds-based"

Reply:
I don't see why using this without xDS is non-trivial.  If the CDS part is just a way of enabling or disabling this, then we wouldn't actually need that at all.  And I think it should be possible to configure the HTTP filter via the gRPC service config without too much trouble.

Reply:
Agree, it would be great to support this without xDS as well - both client side ( cookie jar ) and server side ( setting a cookie) are possible today without changes to the API ( by using metadata ). The only missing piece is an API to 
configure the LB to check the metadata for a cookie and use it. 

Java HttpSession is a very good example - at least for Java it would be great to follow the same pattern. I don't know 
if gRPC is usable with Servlets - but if it is I suspect people may actually use sessions using the servlet session features.

However - I don't know how xDS would be used if first class API would be available.

Reply:
> ...configure the HTTP filter via the gRPC service config without too much trouble.

But http filter support in gRPC is described only in connection with xDS ([A39: xDS HTTP Filter Support](https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md)) so I thought this design being dependent on http filters will also be xDS dependent.

I guess it is possible to abstract out the http filter config as gRPC service config but I would prefer to do that as a separate design since the current need is for xDS based stateful session affinity support.

Reply:
HttpFilter is an artifact of xDS config, for compatibility with Envoy. 

gRPC has interceptor support - so user can plug in a cookie jar if needed without any change in gRPC, and I suspect they already do this to support existing persistent sessions ( many load balancers support them ). 

What is missing in gRPC is a way to specify that the load balancer will use something in metadata to pick the sticky endpoint. For example gRPC uses a number of env variables - the simplest and least intrusive would be to define a
GRPC_ENDPOINT_IP_COOKIE=XYZ, and if it is set the LB would first check if a cookie with this name is present
in the metadata, with  and use it.  

Envoy style of persistent LB is pretty different from everyone else - virtually all others use encrypted cookies, and do the 'persistence' in middle load balancers.

Reply:
I haven't looked recently at the API exposed for load balancing - this is a variant of 'lookaside lb' and may be implementable
already in a hacky way with the existing gRPC code (using grpc/balancer package for example).

Technically the user can also implement the same persistence by extracting the cookie and Dial(IP_FROM_COOKIE), and in case of failure fallback to Dial(service). Pretty sure this is 100% possible today, and I think it has the same behavior - so we may just need to document how an envoy client/server will set the cookies so a gRPC client can interop.

Reply:
This design is limited to the use-case of cookie based session affinity matching the Envoy behavior. And we also have certain time constraints to get this done. Since httpFilter is xDS based I'll leave this also as xDS based

Reply:
Even if we don't want to do any additional work to support configuring the filter via service config, the filter will still be implemented as a stand-alone thing (an interceptor in Java and Go; a filter in C-core and Node) that users could configure independently.  So this really does not need to be strictly an xDS-specific feature.

I'd like input from @ejona86 and @dfawley on this.

Reply:
> And I think it should be possible to configure the HTTP filter via the gRPC service config without too much trouble.

Curious if we have examples of a feature using an http-filter that's configured through service config (and not xds?)

Reply:
- The title is already "A55: xDS-Based Stateful Session Affinity for Proxyless gRPC" 
- the file name is already `A55-xds-stateful-session-affinity.md` 
- renaming the policy name